Add missing method to ChangesetFileReader

rschili opened the pull request:
ChangesetReader::OnConflict creates an arg which is supposed to have a getColumnNames method which is given to `private onChangesetConflict(args: DbMergeChangesetConflictArgs)` on iModelDb.

We correctly construct the arg for LocalChangeset but not for ChangesetFileReader but both invoke the same callback method in iTwinJs.

Must have been an oversight.

Found during my indirect flag testing (changes which we reverted so none of the consuming code is pushed). But I want to merge this fix anyways.

Writing a test for this in the native repo would be a lot of work, so I decided to skip on that. We have no satisfactory test setup for this sort of test.
We can add a test on the iTwin side once this is live. I felt bundling this PR with an iTwin one and having a new Addon build release is a bit drastic. I'll add a test once this has naturally made its way up.

pmconne commented:
> I felt bundling this PR with an iTwin one and having a new Addon build release is a bit drastic. I'll add a test once this has naturally made its way up.

Why? It's easy and automated.

rschili commented:
@pmconne 

> > Why? It's easy and automated.
> 
> Flaky builds

Haha, I wrote Flaky builds and at the same time 2 build steps in this PR failed to set-up node 22.

But let me add some more honest context: I found this missing method during prototyping. I tried to use the method inside IModelDb.onChangesetConflict().
We have one single test on iTwin which mokcs that method, but it doesn't do anything inside, and uses a different code path (MergeChanges vs ChangesetReader). So hitting this method in a test isn't easy.

For rebase we have a way to add handlers (`changeMergeManager.addConflictHandler`, incoming calls go through `TxnManager._onRebaseLocalTxnConflict()`), but for the `IModelDb.onChangesetConflict()` code path, we do not have such a mechanism, meaning I'd have to mock the entire iModelDb API registered against the addon. That seems excessive. We have so many blind spots in our tests in that area, I'm worried to even start going down this hole...

ColinKerr commented:
> > Flaky builds
> 
> @aruniverse when our builds become so flaky that developers are reluctant to employ best practices, we need to fix them. Urgently.

Second this.  

@rschili please at least create the test on the iTwin.js side and link that PR to the issue for this change.  That will add some tracking for the future without tight coupling with the PRs

rschili commented:
@pmconne @ColinKerr Hitting this with a test has been quite a hassle.
When I didn't successfully hit it, I set a breakpoint into the callback-call and ran our entire test suite (native and itwin) with a debugger attached. There is not a single hit.

I did find one skipped test, which mocks the method, but it does that to make sure it does NOT get called. The test was disabled by Paul due to flakiness:
https://github.com/iTwin/itwinjs-core/issues/7730

Fixed a bunch of issues I found along the way, I also did add tests for many other callbacks here in the native addon tests.

I asked the following to @khanaffan in teams but still need an answer. It is quite possible this method is no longer needed at all, as it seems to have been replaced by a rebase variant.

Teams message:
I'm getting very confused about the conflict callbacks on Changeset Apply into ITwinJs.
 
We make the callbacks in iModelCore/iModelPlatform/DgnCore/ChangesetTxns.cpp

- `ChangesetFileReader::_OnConflict` invokes `IModelDb.onChangesetConflict` in iTwin.
- `LocalChangeSet::_OnConflict` invokes `TxnManager._onRebaseLocalTxnConflict` in iTwin.

Both methods have similar signature

```ts
//IModelDb
onChangesetConflict(args: DbMergeChangesetConflictArgs): DbConflictResolution | undefined {}
//TxnManager
_onRebaseLocalTxnConflict(internalArg: DbRebaseChangesetConflictArgs): DbConflictResolution {}
```

The args are given very similar values as they originate from the same sqlite callback. I fixed the first callback by pasting the missing stuff from the second callback.
 
LocalChangeset is exclusively used in DgnCore::TxnManager::PullMergeEnd() Affan added this 8 months ago with the Rebase Pullmerge PR.

To me it looks like our strategy is now always "Rebase"? I thought we had 2 code paths, one for merge and one for rebase.
So is the callback on IModelDb just relic that is not used anymore?

rschili commented:
@khanaffan it appears my change to `` had some unintended side-effect.
Now the assertion here in imodel-native-internal (`m_db->Txns().IsTracking()`) is failing.

```cpp
  void SetupSeedProject(WCharCP inFileName, Db::OpenMode mode, bool wantStandalone) {
    auto seedFileName = GetSeedDbFileName();
    if (Db::OpenMode::ReadWrite == mode)
      m_db = DgnDbTestUtils::OpenSeedDbCopy(seedFileName, inFileName);
    else
      m_db = DgnDbTestUtils::OpenSeedDb(seedFileName);

    if (wantStandalone) {
      SetAsStandaloneDb(m_db, mode);
      ASSERT_TRUE(m_db->IsStandalone());
      ASSERT_TRUE(Db::OpenMode::ReadWrite != mode || m_db->Txns().IsTracking());
    }
```

So seems like we have test code relying in the bug that I fixed.

rschili commented on iModelCore/iModelPlatform/DgnCore/ChangesetTxns.cpp:
```diff
@@ -236,6 +236,15 @@ const auto jsIModelDb = m_dgndb->GetJsIModelDb();
+            arg.Set("getColumnNames", Napi::Function::New(env, [&](const Napi::CallbackInfo&) -> Napi::Value {
+                auto array = Napi::Array::New(env);
+                bvector<Utf8String> columns;
+                if (m_dgndb->GetColumns(columns, iter.GetTableName().c_str())) {
```
> @khanaffan  you can enable getColumns() function if `disableSchemaCheck` is set to `false` or `undefined`. Otherwise you getColumns() should throw exception on typescript side. This is not great check but good enough to catch some mismatch when number of columns do not match.

But there is no code on the typescript side for this. We are constructing the arg in this method, and that's what arrives at the typescript side, there is no wrapper around it, so you cannot prevent people trying to call getColumnNames on the arg.

Any thoughts on my last comment, about imodel-native-internal SetupSeedProject now failing after I fixed `QueryStandaloneEditFlags`? This is an annoying one, because it seems the fix now alters the flag in existing code, which relies on the buggy behavior. So TxnsTracking may now be enabled/disabled in cases where it was the opposite before. Even if this was a bug, it can break consumers.
